docs/tutorials/random-numbers.md

```diff
@@ -1,8 +1,195 @@
+In NumPy, pseudo random number generation is based on a global `state`, which can be set to a deterministic initial condition using {func}`np.random.seed`.
+
+```{code-cell}
+import numpy as np
+np.random.seed(0)
+```
+
+You can inspect the content of the state using the following command.
+
+```{code-cell}
+def print_truncated_random_state():
+  """To avoid spamming the outputs, print only part of the state."""
+  full_random_state = np.random.get_state()
+  print(str(full_random_state)[:460], '...')
+
+print_truncated_random_state()
+```
+
+The `state` is updated by each call to a random function:
+
+```{code-cell}
+np.random.seed(0)
+print_truncated_random_state()
+```
+
+```{code-cell}
+_ = np.random.uniform()
+print_truncated_random_state()
+```
+
+NumPy allows you to sample both individual numbers, or entire vectors of numbers in a single function call. For instance, you may sample a vector of 3 scalars from a uniform distribution by doing:
+
+```{code-cell}
+np.random.seed(0)
+print(np.random.uniform(size=3))
+```
+
+NumPy provides a *sequential equivalent guarantee*, meaning that sampling N numbers in a row individually or sampling a vector of N numbers results in the same pseudo-random sequences:
+
+```{code-cell}
+np.random.seed(0)
+print("individually:", np.stack([np.random.uniform() for _ in range(3)]))
+
+np.random.seed(0)
+print("all at once: ", np.random.uniform(size=3))
+```
+
+## Random numbers in JAX
+
+JAX's random number generation differs from NumPy's in important ways, because NumPy's
+PRNG design makes it hard to simultaneously guarantee a number of desirable properties.
+Specifically, in JAX we want PRNG generation to be:
+
+1. reproducible,
+2. parallelizable,
+3. vectorisable.
+
+We will discuss why in the following. First, we will focus on the implications of a PRNG design based on a global state. Consider the code:
+
+```{code-cell}
+import numpy as np
+
+np.random.seed(0)
+
+def bar(): return np.random.uniform()
+def baz(): return np.random.uniform()
+
+def foo(): return bar() + 2 * baz()
+
+print(foo())
+```
+
+The function `foo` sums two scalars sampled from a uniform distribution.
+
+The output of this code can only satisfy requirement #1 if we assume a predictable order of execution for `bar()` and `baz()`.
+This is not a problem in NumPy, which always evaluates code in the order defined by the Python interpreter.
+In JAX, however, this is more problematic: for efficient execution, we want the JIT compiler to be free to reorder, elide, and fuse various operations in the function we define.
+Further, when executing in multi-device environments, execution efficiency would be hampered by the need for each process to synchronize a global state.
+
+### Explicit random state
+
+To avoid this issue, JAX avoids implicit global random state, and instead tracks state explicitly via a random `key`:
+
+```{code-cell}
+from jax import random
+
+key = random.key(42)
```

Comment:
Also add some kind of temporary disclaimer here mentioning that:
* `key` and special dtypes were recently introduced, and many JAX-related libraries still assume that you use the older `PRNGKey` constructor; and
* for more, see [JEP 9263](https://jax.readthedocs.io/en/latest/jep/9263-typed-keys.html)

?

Reply:
Do you think the note at the top of the document is the right place for that info?

Reply:
I missed that in the markdown view. I should have said: should we move the note down here so that it's local to the constructor usage?

Reply:
Sounds good - I moved that note to just below this.